lib/serialization/compositeSpec.ts

```diff
@@ -129,76 +201,188 @@ export function compositeSpec(typeName: string, propertySpecs: { [propertyName:
+                }
               }
-            }
-          } else {
-            // JSON
-            serializedChildPropertyName = childPropertySpec.serializedName || childPropertyName;
-          }
 
-          // This part is for handling property flattening. If the serialized name has dots in its
-          // name (a.b.c), then we handle those name parts (a, b, c) as individual levels in the
-          // serialized value.
-          const serializedChildPropertyNameParts: string[] = splitSerializeName(serializedChildPropertyName);
-          if (serializedChildPropertyNameParts.length > 1) {
-            for (let i = 0; i < serializedChildPropertyNameParts.length - 1; ++i) {
-              const namePart: string = serializedChildPropertyNameParts[i];
-              if (!serializedPropertyParent[namePart]) {
-                break;
-              } else {
-                serializedPropertyParent = serializedPropertyParent[namePart];
+              const childPropertyPath: PropertyPath = propertyPath.concat([childPropertyName], serializedChildPropertyNameParts);
+
+              // Get the child property's value spec.
+              const childPropertyValueSpec: TypeSpec<any, any> = resolveValueSpec(options, childPropertyPath, childPropertySpec.valueSpec);
+
+              const serializedChildPropertyValue: any = (!serializedPropertyParent ? undefined : serializedPropertyParent[serializedChildPropertyName]);
+              if (serializedChildPropertyValue != undefined) {
+                result[childPropertyName] = childPropertyValueSpec.deserialize(childPropertyPath, serializedChildPropertyValue, options);
+              } else if (childPropertySpec.required && !childPropertySpec.constant) {
+                failDeserializeMissingRequiredPropertyCheck(options, childPropertyPath, childPropertyValueSpec);
               }
             }
+          }
+        }
+      }
+
+      return result;
+    }
+  };
+}
+
+function serializationDiscriminatorPropertyNameGetter(derivedTypes: PolymorphicDerivedTypes): string {
+  return derivedTypes.discriminatorPropertyName;
+}
 
-            if (serializedPropertyParent) {
-              serializedChildPropertyName = serializedChildPropertyNameParts[serializedChildPropertyNameParts.length - 1];
+function deserializationDiscriminatorPropertyNameGetter(derivedTypes: PolymorphicDerivedTypes): string {
+  return derivedTypes.discriminatorPropertySerializedName || derivedTypes.discriminatorPropertyName;
+}
+
+/**
+ * Get all of the CompositeTypeSpecs that the provided value must implement if the starting
+ * CompositeTypeSpec is the provided compositeSpec.
+ */
+function getAllPolymorphicCompositeTypeSpecs(value: CompositeType, propertyPath: PropertyPath, compositeSpec: CompositeTypeSpec, options: SerializationOptions, discriminatorPropertyNameGetter: (derivedTypes: PolymorphicDerivedTypes) => string): CompositeTypeSpec[] {
+  const result: CompositeTypeSpec[] = [];
```

Comment:
What happens if we cannot find a polymorphicCompositeType for a given discriminator value? Do we then serialize or deserialize as that type or throw an Error saying that we could not find a compositeType?

Make sure it is the former as we added support for that in ms-rest for cognitive services.

Reply:
My current implementation is to throw an error indicating that I couldn't find the requested polymorphic CompositeTypeSpec. I'll make this configurable.